kaiax/builder/impl/getter.go

```diff
@@ -100,3 +200,162 @@ func (b *BuilderModule) IsConflict(prevBundles []*builder.Bundle, newBundles []*
 
 	return false
 }
+
+// Filter removes elements at indices specified in `toRemove`.
+func Filter[T any](slice *[]T, toRemove map[int]bool) []T {
+	ret := make([]T, 0)
+	for i := 0; i < len(*slice); i++ {
+		if !toRemove[i] {
+			ret = append(ret, (*slice)[i])
+		}
+	}
+	return ret
+}
+
+func FindBundleIdx(bundles []*builder.Bundle, tx *types.Transaction) int {
+	for i, bundle := range bundles {
+		if bundle.Has(tx.Hash()) {
+			return i
+		}
+	}
+	return -1
+}
+
+func FindBundleIdxAsTxOrGen(bundles []*builder.Bundle, txOrGen interface{}) int {
+	for i, bundle := range bundles {
+		for _, txOrGenInBundle := range bundle.BundleTxs {
+			if EqualTxOrGen(txOrGenInBundle, txOrGen) {
+				return i
+			}
+		}
+	}
+	return -1
+}
+
+func EqualTxOrGen(txOrGenIX, txOrGenIY interface{}) bool {
+	var (
+		txOrGenXHash common.Hash
+		txOrGenYHash common.Hash
+	)
+
+	switch txOrGenX := txOrGenIX.(type) {
+	case *types.Transaction:
+		txOrGenXHash = txOrGenX.Hash()
+	case builder.TxGenerator:
+		txOrGenXHash = txOrGenX.Hash
+	}
+
+	switch txOrGenY := txOrGenIY.(type) {
+	case *types.Transaction:
+		txOrGenYHash = txOrGenY.Hash()
+	case builder.TxGenerator:
+		txOrGenYHash = txOrGenY.Hash
+	}
+
+	return bytes.Equal(txOrGenXHash.Bytes(), txOrGenYHash.Bytes())
+}
+
+func SetCorrectTargetTxHash(bundles []*builder.Bundle, txs []interface{}) []*builder.Bundle {
+	ret := make([]*builder.Bundle, 0)
+	for _, bundle := range bundles {
+		bundle.TargetTxHash = FindTargetTxHash(bundle, txs)
+		ret = append(ret, bundle)
+	}
+	return ret
+}
+
+func FindTargetTxHash(bundle *builder.Bundle, txs []interface{}) common.Hash {
+	// If this is never updated it is expected to return an empty hash.
+	targetTxHash := common.Hash{}
+	for i, tx := range txs {
+		// For index greater than 0, if it can be cast to *types.Transaction then record this as the TargetTxHash.
+		if i > 0 {
+			if txTarget, ok := txs[i-1].(*types.Transaction); ok {
+				targetTxHash = txTarget.Hash()
+			}
+		}
+		// If tx is the first tx in the bundle then there is no need to look further.
+		if EqualTxOrGen(bundle.BundleTxs[0], tx) {
+			break
+		}
+	}
+	return targetTxHash
+}
+
+func ShiftTxs(txs *[]interface{}, num int) {
+	if len(*txs) <= num {
+		*txs = (*txs)[:0]
+		return
+	}
+	*txs = (*txs)[num:]
+}
+
+func PopTxs(txs *[]interface{}, num int, bundles *[]*builder.Bundle, signer types.Signer) {
```

Comment:
It seems the current PopTxs can't handle the case that `num == 1` and `txs[0] == genTx`. The 

```
// Create test bundles
	bundles := []*builder.Bundle{
		{
			BundleTxs: []interface{}{g, txs[1], txs[2]},
		},
	}

	testCases := []struct {
		name            string
		incorporatedTxs []interface{}
		numToPop        int
		bundles         []*builder.Bundle
		expectedTxs     []interface{}
	}{
		{
			name:            "One bundle - first tx is generator",
			incorporatedTxs: []interface{}{g, txs[1], txs[2], txs[3], txs[4], txs[5], txs[6]},
			numToPop:        1,
			bundles:         []*builder.Bundle{bundles[0]},
			expectedTxs:     []interface{}{txs[4], txs[5],txs[6]},
		},
	}
```

Reply:
Ah, bundle[0] is [tx1, tx2], so when numPop=1, only the tx generator is deleted.
So it seems that the expected setting is incorrect.
```
                {
			name:            "One bundle - first tx is generator",
			incorporatedTxs: []interface{}{g, txs[1], txs[2], txs[3], txs[4], txs[5], txs[6]},
			numToPop:        1,
			bundles:         []*builder.Bundle{bundles[0]},
			expectedTxs:     []interface{}{txs[1], txs[2], txs[3], txs[4], txs[5],txs[6]},
		},
```

Reply:
I added a test case for "One bundle - first tx is generator". Thank you.

Reply:
It seems we need to draw edges between generator and txs if they're in the same bundle. So we need to add Lewis' testcase

Reply:
> Ah, bundle[0] is [tx1, tx2], so when numPop=1, only the tx generator is deleted.
So it seems that the expected setting is incorrect.

I overlooked the premise of the bundle.
I got fixed here: 7fa8532